app/src/rgb_underglow.c

```diff
@@ -434,4 +440,67 @@ int zmk_rgb_underglow_change_spd(int direction) {
     return zmk_rgb_underglow_save_state();
 }
 
+#if IS_ENABLED(CONFIG_ZMK_RGB_UNDERGLOW_AUTO_OFF_IDLE)
+static bool auto_off_idle_prev_state = false;
+#endif
+
+#if IS_ENABLED(CONFIG_ZMK_RGB_UNDERGLOW_AUTO_OFF_USB)
+static bool auto_off_usb_prev_state = false;
+#endif
+
+#if IS_ENABLED(CONFIG_ZMK_RGB_UNDERGLOW_AUTO_OFF_IDLE) ||                                          \
+    IS_ENABLED(CONFIG_ZMK_RGB_UNDERGLOW_AUTO_OFF_USB)
+static int rgb_underglow_event_listener(const zmk_event_t *eh) {
+
+#if IS_ENABLED(CONFIG_ZMK_RGB_UNDERGLOW_AUTO_OFF_IDLE)
+    if (as_zmk_activity_state_changed(eh)) {
+        bool new_state = (zmk_activity_get_state() == ZMK_ACTIVITY_ACTIVE);
+        if (state.on == new_state) {
+            return 0;
+        }
+        if (new_state) {
+            state.on = auto_off_idle_prev_state;
+            auto_off_idle_prev_state = false;
+            return zmk_rgb_underglow_on();
+        } else {
+            state.on = false;
+            auto_off_idle_prev_state = true;
+            return zmk_rgb_underglow_off();
+        }
```

Comment:
If we have one common storage, we can move this logic into some new function, e.g.:

```
static void zmk_rgb_underglow_auto_state(bool new_state) {
    if (new_state) {
            state.on = auto_off_prev_state;
            auto_off_prev_state = false;
            return zmk_rgb_underglow_on();
        } else {
            state.on = false;
            auto_off_prev_state = true;
            return zmk_rgb_underglow_off();
        }
}
```

And then the implementation of these two cases in the event handler can be drastically simplified.